### ISSUE TITLE
HTTP/WebSocket Server framework, C++11 header-only using Beast and Boost.Asio

### ISSUE THREAD

**Issue by fffaraz** [fffaraz/awesome-cpp#358]
The Beast library now offers a complete server framework as an example you can build and run: https://github.com/vinniefalco/Beast/tree/master/example/server-framework
Its got these features:
Comes with a WebSocket echo port handler (sync and async)
Comes with an HTTP port handler (sync and async)
Configurable PortHandler concept for user-defined behaviors
Create your own "ports" by writing a class to take ownership of sockets from incoming connections, plug it in to the server framework, and then publish it as open source so other people can use your ports!
The HTTP ports that come with this example are extensible, defining a Service concept which handles HTTP requests. A Service List aggregates Service objects, and the implementation gives each one a chance to handle the request in turn (kind of like a router, but more powerful!) The example comes with two services:
file_service serves files from a path on your system, like a web server
ws_upgrade_service lets you accept WebSocket connections from the same port as your web server and hands them off to the websocket port handler!
As with PortHandler, you can write your own Service types and plug them into this extensible framework, then share them with other people. For example you could implement a simple Troll-box style WebSocket chat Service which plugs right in to the framework and release that as your own open source library!
This is all built on top of Beast to show off the power of the library. People have asked for higher level functionality. You spoke - we listened! Now you can create a richly featured HTTP/WebSocket server in just a matter of seconds by cloning the server-framework example!
Beast is a C++ header-only library serving as a foundation for writing interoperable networking libraries by providing low-level HTTP/1, WebSocket, and networking protocol vocabulary types and algorithms using the consistent asynchronous model of Boost.Asio.
Beast: https://github.com/vinniefalco/Beast
Documentation: http://vinniefalco.github.io/beast/index.html

**Comment by vinniefalco** [comment]
Do I need to do something here? Also, Beast is now part of Boost

**Comment by DragonOsman** [comment]
Is there any documentation for the server framework? If not, I think one should be written so that users can more easily learn how to use it.  And would one have to write a client for it, or would it be fine to use one of the client examples that come with Beast?

**Comment by eXpl0it3r** [comment]
Updated URL: https://github.com/boostorg/beast

I think this should be added to the list!

@DragonOsman This is the awesome-cpp repository, for questions regarding the library, check the libraries' resources on how to get help. 🙂

**Comment by vinniefalco** [comment]
I got rid of that silly framework server example, it was too complicated for the few features it offered. I scaled it back and I have instead created **many** smaller example programs in a variety of styles: synchronous, asynchronous, coroutine, and stackless coroutine, with clients and servers, with or without SSL, for HTTP or WebSocket. There are also "advanced" servers which support both HTTP and WebSocket on the same port, and a "flex" server that can handle plain and SSL on the same port. See:

http://www.boost.org/doc/libs/master/libs/beast/doc/html/beast/examples.html

This many smaller examples serves users better than the previous one large example.
